File: exact_1d.m
% exact solution for 1D model

function exact_1d(xI, kxI, sigmax, init_s, A, mass, fID)
    tic;

    c1 = sqrt(1 - init_s);
    c2 = sqrt(init_s);

    N = 2;
    L = 16;
    M = 256;                    

    dt = 0.1;
    Nstep = 3000; %ceil(7.5 / (k(1) / mass) / dt); % 5000;
    tgraph = 100;

    enable_plot = true;
    enable_plot_diab_surf = false;
    enable_plot_surf = false;
    % grids
    x0 = linspace(-L/2, L/2, M)';
    dx = x0(2) - x0(1);
    dkx = 2 * pi / M / dx;
    kx0 = (-M/2:M/2-1)' * dkx;
    % construct TU on k grid
    T = kx0.^2 / 2 / mass;
    TU = exp(-1i * dt * T);
    TU = fftshift(TU);
    % construct VU
    VU = zeros(M,N,N);
    Hs = zeros(M,N,N);
    for j=1:M
        B = 3.0;
        theta = pi / 2 * (erf(B * x0(j)) + 1);

        H = A * [-cos(theta), sin(theta); sin(theta), cos(theta)];
        VU(j,:,:) = expm(-1i * dt / 2 * H);
        Hs(j,:,:) = H;
    end
    % Initial wavefunction -- Gaussian wavepacket
    psi0 = zeros(M,M,N);
    psi0(:,1) = c1 * exp(1i*(kxI*x0)) .* exp(-(x0-xI).^2/sigmax^2);
    psi0(:,2) = c2 * exp(1i*(kxI*x0)) .* exp(-(x0-xI).^2/sigmax^2);
    psi0 = psi0 / sqrt(sum(sum(sum(abs(psi0).^2))));
    % psim & psi_k_m -- for plot
    psi0_k(:,1) = fftshift(fft(psi0(:,1)));
    psi0_k(:,2) = fftshift(fft(psi0(:,2)));
    psi0_k = psi0_k / sqrt(sum(sum(abs(psi0_k).^2)));
    psi_max = max(max(max(abs(psi0).^2)));
    psi_k_max = max(max(max(abs(psi0_k).^2)));
    % propagate WF
    psi = psi0;
    for t=0:Nstep-1
        % exp(-iVdt/2) * |Psi> in diab repr
        psi_k(:,1) = VU(:,1,1).*psi(:,1) + VU(:,1,2).*psi(:,2);
        psi_k(:,2) = VU(:,2,1).*psi(:,1) + VU(:,2,2).*psi(:,2);
        % exp(-iTdt) * psi
        psi_k(:,1) = TU .* fft(psi_k(:,1));
        psi_k(:,2) = TU .* fft(psi_k(:,2));
        % exp(-iVdt/2) * psi
        psi_k(:,1) = ifft(psi_k(:,1));
        psi_k(:,2) = ifft(psi_k(:,2));
        psi(:,1) = VU(:,1,1).*psi_k(:,1) + VU(:,1,2).*psi_k(:,2);
        psi(:,2) = VU(:,2,1).*psi_k(:,1) + VU(:,2,2).*psi_k(:,2);
        % analysis & report
        if mod(t,tgraph) == 0
            % analysis in diab
            psi_k(:,1) = fftshift(fft(psi(:,1)));
            psi_k(:,2) = fftshift(fft(psi(:,2)));
            psi_k = psi_k / sqrt(sum(sum(abs(psi_k).^2)));
            
            norm_k1 = sum(abs(psi_k(:,1)).^2) + 1e-16;
            norm_k2 = sum(abs(psi_k(:,2)).^2) + 1e-16;

            p1x = sum(abs(psi_k(:,1)).^2 .* kx0) / norm_k1;
            p2x = sum(abs(psi_k(:,2)).^2 .* kx0) / norm_k2;

            KE = sum((abs(psi_k(:,1)).^2 + abs(psi_k(:,2)).^2) .* (kx0.^2) / 2 / mass);
            PE = sum( conj(psi(:,1)) .* Hs(:,1,1) .* psi(:,1) + conj(psi(:,1)) .* Hs(:,1,2) .* psi(:,2) + conj(psi(:,2)) .* Hs(:,2,1) .* psi(:,1) + conj(psi(:,2)) .* Hs(:,2,2) .* psi(:,2) );

            % output
            if t == 0
                fprintf(fID, '# EXACT\n');
                fprintf(fID, '# xI = %8.4f kxI = %8.4f sigmax = %8.4f A = %8.4f init_s = %8.4f c1 = %8.4f c2 = %8.4f \n', ...
                                    xI, kxI, sigmax, A, init_s, c1, c2);
                fprintf(fID, '# L = %8.4f M = %8d dt = %8.4f Nstep = %8d tgraph = %8d\n', ...
                                    L, M, dt, Nstep, tgraph);
                fprintf(fID, '#%9s%16s%16s%16s%16s%16s\n', ...
                                't', 'n0', 'n1', 'px0', 'px1', 'Etot');
            end

            fprintf(fID, '#%16.10f%16.10f%16.10f%16.10f%16.10f%16.10f\n', ...
                        t*dt, ...
                        sum(sum(abs(psi(:,1).^2))), ...
                        sum(sum(abs(psi(:,2).^2))), ...
                        p1x, p2x, ...
                        KE+PE ...
                        );
            % plot
            if enable_plot == true
                subplot(1,2,1);
                plot(x0,abs(psi(:,1)).^2, '-b', ...
                     x0,abs(psi(:,2)).^2, '--r');
                ylim([0, psi_max]);
                title('Real space Diab');

                subplot(1,2,2);
                plot(kx0,abs(psi_k(:,1)).^2, '-b', ...
                     kx0,abs(psi_k(:,2)).^2, '--r');
                ylim([0, psi_k_max]);
                title('Mom space Diab');

                drawnow;
            end

        end
    end
    fprintf(fID, '%16.10f%16.10f%16.10f%16.10f%16.10f\n', ...
            kxI, ...
            sum(sum(abs(psi(:,1).^2))), ...
            sum(sum(abs(psi(:,2).^2))), ...
            p1x, p2x ...
            );
    fprintf('# '); toc;
end

File: exact_1d.cpp
#include <cstdlib>
#include <cmath>
#include <complex>
#include <algorithm>
#include <string>
#include "misc/fmtstring.hpp"
#include "misc/ioer.hpp"
#include "misc/crasher.hpp"
#include "misc/randomer.hpp"
#include "misc/vector.hpp"
#include "misc/timer.hpp"
#include "misc/fft.hpp"
#include "misc/matrixop.hpp"
#include "boost/numeric/odeint.hpp"
#include "boost/math/special_functions/erf.hpp"
#include "boost/program_options.hpp"

using namespace std;
namespace po = boost::program_options;
using boost::numeric::odeint::runge_kutta4;
using boost::math::erf;

double L = 16;
int M = 256;
double mass = 1000.0;

double A = 0.1;
double B = 3.0;

double xI = -3.0;
double sigmax = 1.0;
double kxI = 30.0;

double xwall_left = -99999999.9;
double xwall_right = 99999999.9;

double init_s = 1.0;
int Nstep = 2000;
double dt = 0.1;

int output_step = 100;

string output_mod = "init_s";

inline bool argparse(int argc, char** argv) 
{
    po::options_description desc("Allowed options");
    desc.add_options()
        ("help", "produce help message")
        ("L", po::value<double>(&L), "grid para, the grid is [-L/2, L/2]")
        ("M", po::value<int>(&M), "grid number")
        ("mass", po::value<double>(&mass), "mass")
        ("A", po::value<double>(&A), "potential para")
        ("B", po::value<double>(&B), "potential para")
        ("init_x", po::value<double>(&xI), "init x")
        ("sigma_x", po::value<double>(&sigmax), "init sigma x")
        ("init_px", po::value<double>(&kxI), "init px")
        ("init_s", po::value<double>(&init_s), "init surface")
        ("xwall_left", po::value<double>(&xwall_left), " left boundary x value to check end")
        ("xwall_right", po::value<double>(&xwall_right), " right boundary x value to check end")
        ("Nstep", po::value<int>(&Nstep), "# step")
        ("dt", po::value<double>(&dt), "time step")
        ("output_step", po::value<int>(&output_step), "output step")
        ("output_mod", po::value<string>(&output_mod), "output mode, can be init_s or init_px")
        ;
    po::variables_map vm; 
    po::store(po::parse_command_line(argc, argv, desc), vm);
    po::notify(vm);    
    if (vm.count("help")) {
        std::cout << desc << "\n";
        return false;
    }
    return true;
}

vector< complex<double> > cal_H(const double x) {
    vector< complex<double> > H(4);
    double theta = 0.5 * M_PI * (erf(B * x) + 1.0);
    H[0+0*2] = -cos(theta);
    H[0+1*2] = sin(theta);
    H[1+0*2] = conj(H[0+1*2]);
    H[1+1*2] = cos(theta);
    return A * H;
}

vector< complex<double> > myfftshift(const vector< complex<double> >& Z) 
{
    vector< complex<double> > rst(M);
    for (int j = 0; j < M/2; ++j) {
        rst[j] = Z[j+M/2];
        rst[j+M/2] = Z[j];
    }
    return rst;
}

bool check_end( const vector< complex<double> >& psi0, const vector< complex<double> >& psi1, 
                const vector<double>& xarr)
{
    // if the WF is outside the wall, return true to stop the program
    double n_outside = 0.0;
    for (int j = 0; j < M; ++j) {
        double x = xarr[j];
        if (x < xwall_left or x > xwall_right) {
            n_outside += pow(abs(psi0[j]), 2) + pow(abs(psi1[j]), 2);
        }
    }
    return (n_outside > 0.01);
}

void exact() {
    // para
    double dx, dkx;
    vector<double> xarr = linspace(-L/2, L/2, M, dx);
    dkx = 2 * M_PI / M / dx;
    vector<double> kxarr = (arange(M) - M / 2) * dkx;
    xwall_left = max(xwall_left, -L/2 * 0.9);
    xwall_right = min(xwall_right, L/2 * 0.9);
    double c0 = sqrt(1 - init_s);
    double c1 = sqrt(init_s);
    // construct TU on k grid
    vector< complex<double> > TU(M);
    for (int j = 0; j < M; ++j) {
        double kx = kxarr[j];
        TU[j] = exp(-matrixop::IMAGIZ * dt * (kx*kx) / 2.0 / mass);
    }
    TU = myfftshift(TU);
    // construct VU on x grid
    vector< complex<double> > V00(M), V01(M), V10(M), V11(M);
    vector< complex<double> > evts00(M), evts01(M), evts10(M), evts11(M);
    vector< complex<double> > H00(M), H01(M), H10(M), H11(M);
    for (int j = 0; j < M; ++j) {
        double x = xarr[j];
        vector< complex<double> > H = cal_H(x);
        vector<double> eva;
        vector< complex<double> > evt, evamat;
        matrixop::eigh(H, eva, evt);

        evamat.assign(4, 0.0);
        evamat[0+0*2] = exp(-matrixop::IMAGIZ * dt / 2.0 * eva[0]);
        evamat[1+1*2] = exp(-matrixop::IMAGIZ * dt / 2.0 * eva[1]);
        auto tmp = matrixop::matmatmatC(evt, evamat, evt, 2, 2);
        V00[j] = tmp[0+0*2];
        V01[j] = tmp[0+1*2];
        V10[j] = tmp[1+0*2];
        V11[j] = tmp[1+1*2];
        H00[j] = H[0+0*2];
        H01[j] = H[0+1*2];
        H10[j] = H[1+0*2];
        H11[j] = H[1+1*2];

        // adiab evts
        if (j == 0) {
            const complex<double> phase0 = evt[0+0*2] / abs(evt[0+0*2]);
            evt[0+0*2] *= conj(phase0);
            evt[1+0*2] *= conj(phase0);
            const complex<double> phase1 = evt[1+1*2] / abs(evt[1+1*2]);
            evt[0+1*2] *= conj(phase1); 
            evt[1+1*2] *= conj(phase1);
        }
        else {
            const complex<double> phase0 = conj(evts00[j-1]) * evt[0+0*2] + conj(evts10[j-1]) * evt[1+0*2];
            evt[0+0*2] *= conj(phase0);
            evt[1+0*2] *= conj(phase0);
            const complex<double> phase1 = conj(evts01[j-1]) * evt[0+1*2] + conj(evts11[j-1]) * evt[1+1*2];
            evt[0+1*2] *= conj(phase1);
            evt[1+1*2] *= conj(phase1);
        }
        evts00[j] = evt[0+0*2];
        evts01[j] = evt[0+1*2];
        evts10[j] = evt[1+0*2];
        evts11[j] = evt[1+1*2];
    }
    // initialized WF
    /*
    vector< complex<double> > psi0(M, 0.0), psi1(M, 0.0);
    for (int j = 0; j < M; ++j) {
        double x = xarr[j];
        psi0[j] = c0 * exp(matrixop::IMAGIZ * (kxI * x)) * exp(-pow((x - xI) / sigmax, 2));
        psi1[j] = c1 * exp(matrixop::IMAGIZ * (kxI * x)) * exp(-pow((x - xI) / sigmax, 2));
    }
    double nm = norm(psi0 | psi1);
    psi0 /= nm;
    psi1 /= nm;
    */
    // initialized WF on adiab
    vector< complex<double> > psi0(M, 0.0), psi1(M, 0.0);
    vector< complex<double> > psiad0(M, 0.0), psiad1(M, 0.0);
    for (int j = 0; j < M; ++j) {
        double x = xarr[j];
        psiad0[j] = c0 * exp(matrixop::IMAGIZ * (kxI * x)) * exp(-pow((x - xI) / sigmax, 2));
        psiad1[j] = c1 * exp(matrixop::IMAGIZ * (kxI * x)) * exp(-pow((x - xI) / sigmax, 2));
    }
    double nm = norm(psiad0 | psiad1);
    psiad0 /= nm;
    psiad1 /= nm;
    psi0 = evts00 * psiad0 + evts01 * psiad1;
    psi1 = evts10 * psiad0 + evts11 * psiad1;

    // covinience vairables
    vector<int> dim{ M };
    // statistics
    double KE0 = 0.0, KE1 = 0.0, PE = 0.0;
    double n0trans = 0.0, n0refl = 0.0, n1trans = 0.0, n1refl = 0.0;
    double px0trans = 0.0, px0refl = 0.0, px1trans = 0.0, px1refl = 0.0;
    // propagate WF
    for (int istep = 0; istep < Nstep; ++istep) {
        // exp(-iVdt/2)
        auto psi_k0 = V00 * psi0 + V01 * psi1;
        auto psi_k1 = V10 * psi0 + V11 * psi1;
        // exp(-iTdt)
        psi_k0 = misc::fftn(psi_k0, dim);
        psi_k1 = misc::fftn(psi_k1, dim);
        psi_k0 *= TU;
        psi_k1 *= TU;
        // exp(-iVdt/2)
        psi_k0 = misc::ifftn(psi_k0, dim);
        psi_k1 = misc::ifftn(psi_k1, dim);
        psi0 = V00 * psi_k0 + V01 * psi_k1;
        psi1 = V10 * psi_k0 + V11 * psi_k1;
        // analysis & output
        if (istep % output_step == 0) {
            // get psi_k
            psi_k0 = myfftshift(misc::fftn(psi0, dim));
            psi_k1 = myfftshift(misc::fftn(psi1, dim));
            double nm = norm(psi_k0 | psi_k1);
            psi_k0 /= nm;
            psi_k1 /= nm;
            KE0 = KE1 = PE = 0.0;
            n0trans = n0refl = n1trans = n1refl = 0.0;
            px0trans = px0refl = px1trans = px1refl = 0.0;
            for (int j = 0; j < M; ++j) {
                if (kxarr[j] >= 0.0) {
                    n0trans += pow(abs(psi_k0[j]), 2);
                    n1trans += pow(abs(psi_k1[j]), 2);
                    px0trans += pow(abs(psi_k0[j]), 2) * kxarr[j];
                    px1trans += pow(abs(psi_k1[j]), 2) * kxarr[j];
                }
                else {
                    n0refl += pow(abs(psi_k0[j]), 2);
                    n1refl += pow(abs(psi_k1[j]), 2);
                    px0refl += pow(abs(psi_k0[j]), 2) * kxarr[j];
                    px1refl += pow(abs(psi_k1[j]), 2) * kxarr[j];
                }
                KE0 += pow(abs(psi_k0[j]), 2) * (kxarr[j]*kxarr[j]) / 2.0 / mass;
                KE1 += pow(abs(psi_k1[j]), 2) * (kxarr[j]*kxarr[j]) / 2.0 / mass;
            }
            px0trans /= (n0trans + 1e-16);
            px1trans /= (n1trans + 1e-16);
            px0refl /= (n0refl + 1e-16);
            px1refl /= (n1refl + 1e-16);
            PE = real(sum(conj(psi0) * H00 * psi0 + conj(psi0) * H01 * psi1 + conj(psi1) * H10 * psi0 + conj(psi1) * H11 * psi1));
            // output
            if (istep == 0) {
                ioer::info("# EXACT 2D DIAB OUTPUT");
                ioer::info("# para: ", " L = ", L, " M = ", M, " mass = ", mass, " A = ", A, " B = ", B, 
                                       " xI = ", xI, " sigmax = ", sigmax, " kxI = ", kxI," init_s = ", init_s, " c0 = ", c0, " c1 = ", c1,
                                       " Nstep = ", Nstep, " dt = ", dt, " output_step = ", output_step);
                ioer::info("# dx = ", dx, " dkx = ", dkx, " xwall_left = ", xwall_left, " xwall_right = ", xwall_right);
                ioer::tabout('#', "t", "n0trans", "n0refl", "n1trans", "n1refl", "px0trans", "px0refl", "px1trans", "px1refl", "Etot");
            }
            ioer::tabout('#', istep * dt, n0trans, n0refl, n1trans, n1refl, px0trans, px0refl, px1trans, px1refl, KE0 + KE1 + PE);
            // check end
            if (check_end(psi0, psi1, xarr) == true) {
                ioer::info("# check_end returns true");
                break;
            }
        }
    }
    // final output
    if (output_mod == "init_s") {
        ioer::tabout(init_s, n0trans, n0refl, n1trans, n1refl, px0trans, px0refl, px1trans, px1refl);
    }
    else if (output_mod == "init_px") {
        ioer::tabout(kxI, n0trans, n0refl, n1trans, n1refl, px0trans, px0refl, px1trans, px1refl);
    }
    else {
    }
}

int main(int argc, char** argv) {
    if (argparse(argc, argv) == false) {
        return 0;
    }
    randomer::seed(0);
    timer::tic();
    exact();
    ioer::info("# ", timer::toc());
    return 0;
}
